bangumi/downloader/pikpak.py

```diff
@@ -0,0 +1,64 @@
+import logging
+from typing import List
+
+from pikpakapi import PikPakApi, PikpakException, PikpakAccessTokenExpireException
+from bangumi.entitiy import DownloadItem
+
+from .downloader import Downloader
+
+logger = logging.getLogger(__name__)
+
+
+class PikpakDownloader(Downloader):
+    def __init__(self, **kwargs):
+        super().__init__()
+
+        username = kwargs.pop("username", "")
+        password = kwargs.pop("password", "")
+
+        logger.info(f"Pikpak initializing...")
+
+        self.client = PikPakApi(username, password)
+
+    def connect(self):
+        try:
+            self.client.login()
+            logger.info("Logined to pikpak")
+        except Exception as e:
+            logger.error(f"Pikpak failed to login: {e}")
+
+    def add_torrent_by_magnet(self, magnet: str) -> bool:
+        try:
+            self.client.offline_download(magnet)
+        except PikpakAccessTokenExpireException:
+            self.client.refresh_access_token()
+            try:
+                self.client.offline_download(magnet)
+            except PikpakException as e:
+                logger.error(f"Pikpak failed to download {magnet} with error {e}")
+                return False
+        return True
+
+    def add_torrent_by_url(self, url: str) -> bool:
+        try:
+            self.client.offline_download(url)
+        except PikpakAccessTokenExpireException:
+            self.client.refresh_access_token()
+            try:
+                self.client.offline_download(url)
+            except PikpakException as e:
+                logger.error(f"Pikpak failed to download {url} with error {e}")
+                return False
+        return True
```

Comment:
the code here is redundant (no.31-40 & no.43-52) since the client can identify whether is an HTTP link or a magnet link by itself.
In my opinion, you may extract this block of code to a new function, and call the new function in these two methods.

Reply:
the problem is the client deligates the HTTP link to `add_torrent_by_file`, however, pikpak does not support torrent file but torrent link (in http protocal), thus, I have to override and handle the `add_torrent_by_url` function to make it work.

Reply:
I resolved such redundance by introducing a `handle_api_error` func following the peer implementation. see that whether it is good for you.